Ranks and Stats and Events and End my life.

Kantanomo opened the pull request.

Commit:
Started Stats and Gamestate callbacks

Created the base for stats, got playlist verification CURL working. Working on playlist uploading next

Hooked the function that evaluates the current Gamestate and created a system for adding callback functions that can be ran when the gamestate is changed.

Commit:
More Examples of callbacks

Commit:
Stats Json Creation

Started work on the Stats JSON Creation

Added some functions to Util
ToNarrow: flattens a wchar_t to char*
wcstombs2r:  same as wcstombs2 but instead of a output parameter it returns the output.

Commit:
Update StatsHandler.cpp

Medals and Weapons

Commit:
stats

stats

Commit:
Stats and Ranks

Created custom packet type s_rank_change for allowing the server to set the clients rank.

Made a function in H2MOD to set the local player rank.

Added Playlist uploading, stats uploading and getPlaylistRanks for StatsHandler.

Add some helper functions to Utils for converting int types to different forms of string representation
IntoToString and IntToWString

Commit:
Ranks and Networking

Reworked getPlayerRanks to utilize NetworkSession info as it populates before a game ends.

Also added basic filtering to the getPlayerRanks to not constantly spam clients with packets, and only players who join. also a force all option for when GameState reaches Lobby so that every player is updated.

Issues with this implimentation are that when a player leaves in the lobby and rejoins it will not send them their rank until a match has been played.

Added a new case to Custom packets::handle_channel_message_hook for peer_properties. This is used to leverage the ability to send a joining player thier rank as the Client requests peer properties after it has fully established itself inside a lobby. (Open to better options, but this is the best i've found).

Commit:
Networking Events Statistics

Created event handling for Player Leave/Join from CustomPackets.cpp and H2MOD.cpp

Can register a callback to them which will pass a XUID parameter through it, can probably be changed in the future to have a copy of the player data or point to the NetworkSession::GetPlayerInformation

Fix an issue with rank's that if a player leaves and rejoins in lobby the rank will not appear until a match has been played.

Added a helper function to NetworkSession getPeerXUID which gets the XUID from the peer index.

Commit:
Ranks Ranks Ranks

Patched the call to BuildPostGameCarnageReport to fix the ranks on the Clients side. In the Post Game Carnage report so they are now visible at the end of the game.

Fixed some Datatypes in the PlayerStruct for player_displayed_skill and player_overall_skill.

Commit:
Fuck

Commit:
Revert "Fuck"

This reverts commit cdce707bc62e3571107f1744f0ac30440198797a.

Commit:
Revert "Ranks Ranks Ranks"

This reverts commit 0caf4a653343ac1e77950444e38a170b7181bbb6.

Commit:
Reversion and self deprication

Reverted code back to later commit, because for some reason it started crashing after nothing was changed.

Reorganized the packets handling, player ranks now appear the second the player joins.

Commit:
Some tweeks

Commit:
Stats and handlers

Moved stats handlers for ranks to server only, and fixed the way TeamPlay is encoded.

Commit:
Threadding Events

Added threadding to the callbacks, as noticing if the webserver did not respond in a fast manner for fetching ranks it would cause players to not appear instantaneously.

Commit:
reorganizing.

Commiting before making some big changes.

Commit:
EventHandler

Added EventHandler namespace which is used for what the name implies.

Moved all the events that were added to H2MOD.cpp to this new namespace.

Commit:
Minimum Player Start and a small rank fix

Added a config setting for servers minimum_player_start which allows server owners to stop their matches from starting until the playercount reachs the value provided.

fix an issue with first match ranks not showing.

Commit:
some tweaks

Removed uncessary log call and turned verbose logging off.

pnill commented:
Get it compiling first.

Kantanomo committed:
Merge branch 'development' into development

shayaantx commented on xlive/Project_Cartographer.vcxproj:
```diff
@@ -64,26 +64,26 @@
     <IncludePath>C:\Program Files %28x86%29\Microsoft DirectX SDK %28June 2010%29\Include;.\;..\protobuf-2.6.1;$(IncludePath)</IncludePath>
-    <LibraryPath>..\Directx SDK\lib;..\protobuf-2.6.1\vsprojects\Debug;.\libs;C:\Program Files (x86)\Microsoft DirectX SDK (June 2010)\Lib\x86;C:\Program Files (x86)\Microsoft SDKs\Windows\v7.1A\Lib;C:\Program Files (x86)\Microsoft Visual Studio\Shared\14.0\VC\lib;$(LibraryPath)</LibraryPath>
-    <IncludePath>..\Directx SDK\include;..\protobuf-2.6.1;.\;.\google;.\teamlog;.\teamspeak;C:\Program Files (x86)\Microsoft DirectX SDK (June 2010)\Include;C:\Program Files (x86)\Microsoft SDKs\Windows\v7.1A\Include;C:\Program Files (x86)\Microsoft Visual Studio\Shared\14.0\VC\include;$(IncludePath)</IncludePath>
+    <LibraryPath>F:\Users\Mickenzie\Documents\Development\DirectX\Lib\x86;..\protobuf-2.6.1\vsprojects\Debug;.\libs;C:\Program Files (x86)\Microsoft DirectX SDK (June 2010)\Lib\x86;C:\Program Files (x86)\Microsoft SDKs\Windows\v7.1A\Lib;C:\Program Files (x86)\Microsoft Visual Studio\Shared\14.0\VC\lib;$(LibraryPath)</LibraryPath>
```
no custom user changes

Kantanomo replied:
Are these not supposed to be in .gitignore? Not sure how to resolve this one

shayaantx commented on xlive/H2MOD/Modules/Stats/StatsHandler.h:
```diff
@@ -0,0 +1,105 @@
+		{
+			auto document = getPlayerRanks(forceAll);
+			if (document.MemberCount() == 0)
+			{
```
empty if block?

shayaantx commented on xlive/H2MOD/Modules/Stats/StatsHandler.h:
```diff
@@ -0,0 +1,105 @@
+			{
+				for (auto i = 0; i < document.MemberCount(); i++)
+				{
+					//LOG_INFO_GAME(document[i]["XUID"].GetString());
```
remove commented out code or use it

shayaantx commented on xlive/H2MOD/Modules/Stats/StatsHandler.h:
```diff
@@ -0,0 +1,105 @@
+					{
+						if(uploadStats(filepath) == 200)
+						{
+							//Stats upload success
```
empty if block?

shayaantx commented on xlive/H2MOD/Modules/Stats/StatsHandler.cpp:
```diff
@@ -0,0 +1,686 @@
+		LOG_INFO_GAME(L"[H2MOD] playlistVerified loaded playlist is {0}", playlist_file);
+
+	std::string checksum = getChecksum();
+	if (checksum == "")
```
use .empty() instead and return an empty string getChecksum instead of null

shayaantx commented on xlive/H2MOD/Modules/Networking/CustomPackets/CustomPackets.cpp:
```diff
@@ -245,6 +275,21 @@ void __stdcall handle_channel_message_hook(void *thisx, int network_channel_inde
 
 	p_handle_channel_message(thisx, network_channel_index, message_type, dynamic_data_size, packet);
+
+	switch(message_type)
```
why use a switch when you're only check for one case? just use an if

shayaantx commented on xlive/H2MOD/Modules/EventHandler/EventHandler.h:
```diff
@@ -0,0 +1,200 @@
+	void executeNetworkPlayerRemoveCallbacks(int peerIndex);
+
+	///////////////////////////////////////////////
+	//Someone smarter than me can figure this out//
```
no commented out code, if someone else is supposed to finish this, then this shouldn't be up for PR yet

Kantanomo replied:
Future me is smarter than me, it was meant for myself. I can remove it I guess.

shayaantx commented on xlive/H2MOD/Modules/Config/Config.cpp:
```diff
@@ -721,6 +729,8 @@ void ReadH2Config() {
 
 				H2Config_additional_pcr_time = ini.GetLongValue(H2ConfigVersionSection.c_str(), "additional_pcr_time", H2Config_additional_pcr_time);
+        
+        H2Config_minimum_player_start = ini.GetLongValue(H2ConfigVersionSection.c_str(), "minimum_player_start", H2Config_minimum_player_start);
```
indenting

shayaantx commented on xlive/H2MOD/Modules/Config/Config.cpp:
```diff
@@ -468,8 +475,10 @@ void SaveH2Config() {
 
 			ini.SetValue(H2ConfigVersionSection.c_str(), "server_playlist", H2Config_dedi_server_playlist);
 
+      ini.SetLongValue(H2ConfigVersionSection.c_str(), "minimum_player_start", H2Config_minimum_player_start);
```
indenting

shayaantx commented on xlive/H2MOD/Modules/Config/Config.cpp:
```diff
@@ -364,7 +365,13 @@ void SaveH2Config() {
 				"\n# Now you have the possibility to change it to your preference."
 				"\n\n"
+        
+        "# minimum_player_start options (Server):"
```
indenting

shayaantx commented on xlive/H2MOD.cpp:
```diff
@@ -1325,6 +1428,16 @@ void H2MOD::ApplyHooks() {
+	//Hook to do stuff after Game State Change
+	p_EvaulateGameState = h2mod->GetAddress<ChangeGameState>(0x1d7738, 0x1BCDA8);
+	PatchCall(h2mod->GetAddress(0x1AD84D, 0x1A67CA), EvaluateGameState);
+	
```
unnecessary space below

shayaantx commented on xlive/H2MOD.cpp:
```diff
@@ -1316,6 +1337,88 @@ void H2MOD::ApplyUnitHooks()
+		int playerCount = NetworkSession::getPlayerCount();
+		if(previousPlayerCount != playerCount)
+		{
+
```
empty if block?

shayaantx commented on xlive/H2MOD.cpp:
```diff
@@ -1316,6 +1337,88 @@ void H2MOD::ApplyUnitHooks()
+			result = 0;
+	} 
+	else
+		result = p_StartCountdownTimer(1, countdown_time, a2, a3, a4);
```
why even have a result variable, just return on each line instead (where you set the variable)

shayaantx commented on xlive/H2MOD/Modules/Stats/StatsHandler.cpp:
```diff
@@ -0,0 +1,686 @@
+		if (curlResult != CURLE_OK)
+		{
+			LOG_ERROR_GAME(L"[H2MOD]::getPlayerRanks failed to execute curl");
+			//LOG_ERROR_GAME(curl_easy_strerror(curlResult));
```
no commented out code (either wrap in verbose check/debug log)

shayaantx commented on xlive/H2MOD/Modules/Stats/StatsHandler.cpp:
```diff
@@ -0,0 +1,686 @@
+}
+
+
+struct stringMe {
```
please come up with reasonable struct names, not "stringMe"

shayaantx commented on xlive/H2MOD/Modules/Stats/StatsHandler.cpp:
```diff
@@ -0,0 +1,686 @@
+	WValue Players(rapidjson::kArrayType);
+	for(auto i = 0; i < 16; i++)
+	{
+		int calcBaseOffset  = baseOffset + (i * 0x94);
```
please break up this method into smaller methods for future development sanity.,...

shayaantx commented on xlive/H2MOD/Modules/Stats/StatsHandler.cpp:
```diff
@@ -0,0 +1,686 @@
+	curl_mime_data(field, "GameStats", CURL_ZERO_TERMINATED);
+	headerlist = curl_slist_append(headerlist, buf);
+
+	curl_easy_setopt(curl, CURLOPT_URL, "https://www.halo2pc.com/test-pages/CartoStat/API/post.php");
```
your implementation is not secure

Kantanomo replied:
Added JWT Token Authentication and made it that servers have to manually be enabled for collection.

shayaantx commented on xlive/H2MOD/Modules/Stats/StatsHandler.cpp:
```diff
@@ -0,0 +1,686 @@
+#endif
+
+static const bool verbose = false;
+StatsHandler::StatsHandler()
```
you don't need an empty constructor

Commit:
Making some requested changes.

Commit:
undone one change.

Commit:
Changed where the initial file json is placed

Moved the path where the json is placed to the root directory (Which is deleted as soon as it's processed)

nukeulater commented on xlive/H2MOD/Modules/Stats/StatsHandler.cpp:
```diff
@@ -592,6 +592,30 @@ void StatsHandler::playerJoinEvent(XUID playerXUID)
+				if (std::wstring(gt) == std::wstring(tgt))
+				{
+					auto rankAddress = h2mod->GetAddress(0x46EF70 + (j * 0x110));
+					WriteValue(rankAddress, WORD(rank));
```
if that memory is not read-only, consider replacing WriteValue with `*h2mod->GetAddress<WORD*>(address) = rank`

nukeulater replied:
WriteValue API should only be used when you are dealing with read-only/execute-only memory.

Kantanomo replied:
Function was removed, as it was deemed unessecary and visually unappealing.

NukeULater commented on xlive/H2MOD/Modules/Stats/StatsHandler.cpp:
```diff
@@ -592,6 +592,30 @@ void StatsHandler::playerJoinEvent(XUID playerXUID)
 	sendRankChange();
 }
+
+static BYTE EmptyByte[2]{ 0, 0 };
```
Why don't declare this inside the FixClientPCRRanks() function stack and not static as well? seems to be used only there

nukeulater commented on xlive/H2MOD/Modules/Networking/NetworkSession/NetworkSession.cpp:
```diff
@@ -52,16 +52,36 @@ int NetworkSession::getPeerCount()
 }
+IN_ADDR NetworkSession::getLocalNetworkAddress()
+{
+	return getCurrentNetworkSession()->membership.peer_info[getLocalPeerIndex()].address.inaOnline;
```
Avoid using this, it'll be encrypted.

nukeulater commented on xlive/H2MOD/Modules/Stats/StatsHandler.cpp:
```diff
@@ -0,0 +1,684 @@
+		WValue Player(rapidjson::kObjectType);
+		
+		
+		auto Gamertag = h2mod->GetAddress<wchar_t*>(0, calcBaseOffset + 0xA);
```
this should be PlayerProperties struct. check https://github.com/pnill/cartographer/blob/development/xlive/Blam/Engine/Players/Players.h#L158

Kantanomo replied:
Cannot use the PlayerProperties, I have tested with it and it does not store player data after they leave.

nukeulater commented on xlive/H2MOD/Modules/EventHandler/EventHandler.h:
```diff
@@ -0,0 +1,116 @@
+	player_join,
+	gamestate_change,
+};
+enum GameState : BYTE
```
This is already defined as https://github.com/pnill/cartographer/blob/development/xlive/Blam/Common/Common.h#L50

nukeulater replied:
Please use that instead

Kantanomo replied:
done

Commit:
Token's

Adding Token Authentication

Commit:
Bugs, Auth, and resolving comments

Fixed a bug that caused the AuthKey to not save correctly after cycling the server

Removed duplicate enum for game_life_cycle

Removed sending Server information with the JSON it's now all pulled using the Token information on the webserver

Removed FixClientPCRRanks

Commit:
removing unecessary code.

Kantanomo committed:
Merge branch 'development' into development

iSupersniper commented:
Ranks Я Ready

Commit:
Regorganizing and VIP Locking

Shifted some code around to make it more readable

Added an optional setting that will lock the server into VIP mode while the game is running. It will allow only the players who were in the lobby in the game start to rejoin when the game is going. Incase there are any network issues.

Added another function to ServerConsole, ServerConsole::SendCommand2 that for runs H2Server/H2Admin commands that is more usable than the eixsting one (SendCommand)

Commit:
Data and Results

Removed a some of arbitrary logging that was used for debugging.

Changed the server result for ServerRegistration to return which features are enabled for the server

Fixed a data type in Players.h for player_displayed_skill it was incorrectly set as a handicap value when it's a byte.

Kantanomo committed:
Merge branch 'development' into development

Commit:
Changed up the verification logic

Simplified the logic for verifying a player has recieved and set their rank.

Commit:
Visual Changes and Hills

Added the visual changes for Spartan specular brightness and the shader template for the visor.

Hooked the function that is called to get the next hill index for KotH, changed the functionality to make it sequential rather than random every time.

Commit:
Fixed visor change and implimented fix for simple INI

Was changing the wrong stem field for the visor fix.

Created a fix for accounts files with the new simpleini class. Migrates the old file to the new format.

Commit:
stuff

Removed uneeded logging

nukeulater committed:
Merge branch 'development' into development

NukeULater commented:
Revert this [commit](https://github.com/pnill/cartographer/pull/280/commits/008664ba8cd82b34ed6959e42b4318575d798682)

Commit:
Wine fixes and some other changes

Changed some calls around to fix issues with StatsHandler when it is running under wine. Really don't know or understand what the issue reall was but the changes fix it. My current best guess is something with pointers or their respective data being overridden.

Removed uncessecary code in Accounts.cpp that was left over from my testing on how to recover old account files.

Commented out code to fix the visor, as I was told not to change shader templates. Leaving it there incase it's deemed okay to do in the future.

Commit:
Party Mode and other changes

Added a feature that will keep parties together when they join a lobby from a custom match.

Added some helper functions to NetworkSession.

Changed the Stats upload to account for all Damage Reporting Types and discard sending Types that have no data.

Added an event handler for RunLoop that I ended up not using that allows you to add functions to main_game_loop_hook without bloating the main funciton.

Commit:
Commands, Events and Invalidation

Created an enum ServerConsole::ServerConsoleCommands that is used in conjuction with the dedi command hook.

Created a new event handler for ServerConsoleCommand that allows callbacks to be executed when a specific server command is ran.

Moved all the event registration for StatsHandler into itself and out of H2MOD.cpp.

Added a new EventHandler to statshandler for detecting when the server is told to skip the match. Which will invalidate the match and stop it from submitting it to the API.

Commit:
Logging changes

Moved some things from LOG_INFO to LOG_TRACE

Commit:
KotH Behaviour Setting

Added a setting that will allow servers to choose the behaviour of KotH, from random hills to sequential hills.

Commit:
KotH Fix

Forgot to add in the server address for HillCount.

Commit:
removed uneeded codes and comments.

iSupersniper committed:
Merge branch 'development' into development
